src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqTouchstoneRepository.kt

```diff
@@ -191,14 +191,75 @@ class JooqTouchstoneRepository(
         return records.map { mapCoverageSet(it) }
     }
 
+    private fun coverageDimensions() : Array<Field<*>>
+    {
+        //The columns in the Coverage table which the target and coverage values are grouped by
+        return arrayOf(COVERAGE.COVERAGE_SET,
+                COVERAGE.YEAR,
+                COVERAGE.COUNTRY,
+                COVERAGE.AGE_FROM,
+                COVERAGE.AGE_TO,
+                COVERAGE.AGE_RANGE_VERBATIM,
+                COVERAGE.GAVI_SUPPORT,
+                COVERAGE.GENDER)
+    }
+
+    private fun validTargetOrNull() : Field<BigDecimal?>
+    {
+        //Coverage and target values only make sense if both are provided. Return the target value of a row providing
+        //coverage is not null, else return null
+        return `when`(COVERAGE.COVERAGE_.isNull(), COVERAGE.COVERAGE_) //This is actually the easiest way to get a known NULL here!
+                .otherwise(COVERAGE.TARGET)
+    }
+
+    private fun validCoverageOrNull() : Field<BigDecimal?>
+    {
+        //Coverage and target values only make sense if both are provided. Return the coverage value of a row providing
+        //target is not null, else return null
+        return `when`(COVERAGE.TARGET.isNull(), COVERAGE.TARGET) //This is actually the easiest way to get a known NULL here!
+                .otherwise(COVERAGE.COVERAGE_)
+    }
+
+    private fun aggregatedValues() : List<Field<*>>
+    {
+        return arrayOf(
+                //Aggregated coverage - the sum of each row's coverage * target (to get total no of fvp's across campaign)
+                //divided by the the total target population
+
+                //This is rounded to 2 dec places if required
+                `when`(trunc(aggregatedCoverage(),2).eq(aggregatedCoverage()), aggregatedCoverage())
+                        .otherwise(round(aggregatedCoverage(),2)).`as`("coverage"),
+                //Aggregated target
+                `when`(trunc(aggregatedTarget(),2).eq(aggregatedTarget()), aggregatedTarget())
+                        .otherwise(round(aggregatedTarget(),2)).`as`("target")
+        ).toList()
+    }
+
+    private fun aggregatedCoverage() : Field<BigDecimal?>
+    {
+        //Danger of divide by zero error here - treat as NULL if summed target is 0
```

Comment:
The rounding here was necessary because I think anytime you put in a div, Postgres increases the precision of the resulting number, even if it doesn't need to. So I was ending up with 0.2500000000000 even after my trunc comparison since 0.25 = 0.2500000000000. Probably a corner case, maybe not worth it. What this does mean is that aggregated values across groups of >1 will have a precision of 2, so might have 1 or 2 trailing zeroes (e.g. 0.70). This is probably OK since subnational is a rarity and won't affect majority of rows.